Ability to remove a key/value modifier. Fixes #5

imalitavakoli opened the pull request:
Now you can easily add and also remove a key/value modifier.
If you set the value of the modifier an empty String, then the modifier
will be removed from the block/element:
`$('.t24-button').setMod('t24-button', 'theme', '');`

imalitavakoli committed:
Ability to remove a key/value. Fixes #5

Now you can easily add and also remove a key/value modifier.
If you set the value of the modifier an empty String, then the modifier
will be removed from the block/element:
`$('.24-button').setMod('24-button', 'primary', '');`

imalitavakoli commented:
I hope this PR can help this great library get more awesome :smile: 

Now if you set the value of a key/value modifier an empty String, then the modifier
will be removed from the block/element:
`$('.24-button').setMod('24-button', 'theme', '');`

And if we set a Event handler for our modifier, just like when removing a Boolean modifier which causes `data.modVal` to get `false`, now setting an empty String for a key/value modifier causes the  `data.modVal` to get `false` too after the modifier removal.

``` javascript
// Set our syntax first
$.BEMsyntax({
    elem: '__',
    modBefore: '_',
    modKeyVal: '_'
});

// Set a Event handler for our '24-button_theme' key/value modifier
$(document).on('setMod:24-button_theme_*', function(e, data) {
    console.log('Block name:', data.block);
    console.log('Element name:', data.elem);
    console.log('Modifier name:', data.modName);
    console.log('Modifier value:', data.modVal);
});

// Now let's set the modifier with 'forest' value
$('.24-button').setMod('24-button', 'theme', 'forest');

/* Console output:

>  Block name: 24-button
>  Element name: undefined
>  Modifier name: theme
>  Modifier value: forest

*/

// And here let's remove the modifier totally by setting its value to ''
$('.24-button').setMod('24-button', 'theme', '');

/* Console output:

>  Block name: 24-button
>  Element name: undefined
>  Modifier name: theme
>  Modifier value: false

*/
```

imalitavakoli committed:
Update ReadMe to mention key/value modifier removal

ingdir commented:
Hi @imalitavakoli , thank you for the contribution! I have several comments though:

1. Could you please add test coverage for your new method to the test file according with the existing tests?
2. I'd prefer if you always use a full `if (...) {...} else {...}` syntax with curly brackets, rather than a reduced `if (...) ... else ...` form, to be in line with what already is in the code.
3. I don't think you have to remove the class name [here](https://github.com/imalitavakoli/jquery-bemhelpers/blob/78c3937d75d449565c953638b8323cdd3e76a45b/jquery.bemhelpers.js#L102), as the previous loop that iterates over all the classnames already takes care of that (I think — did I miss smth?). This also means you can simplify the condition checks there.
4. Maybe the doc needs a more extensive usage example — as with the other methods (you almost have it in the comments but not in the doc)
5. I don't particularly like the idea of resetting the `modVal` to `false` [here](https://github.com/imalitavakoli/jquery-bemhelpers/blob/78c3937d75d449565c953638b8323cdd3e76a45b/jquery.bemhelpers.js#L107) as it makes impossible for the event listeners to understand what has just happened (also, it's quite likely that whoever is listening to a key/value modifier change, always expects the `modVal` to be a `String` instance, not a `Boolean`).
6. Please split long comments into several lines.

I'd really appreciate if you could introduce the above mentioned changes. I'm open for questions/conversation.

ingdir commented on jquery.bemhelpers.js:
```diff
@@ -62,40 +62,50 @@
-                    $this.addClass(classPattern + BEMsyntax.modKeyVal + modVal);
+
+                    // Check whether we should remove the modifer class or not according to 'removeTheAlreadySetValue'
+                    if (removeTheAlreadySetValue === true) $this.removeClass(classPattern + BEMsyntax.modKeyVal + modVal);
```
I'd prefer if you always use a full if (...) {...} else {...} syntax with curly brackets, rather than a reduced if (...) ... else ... form.

ingdir replied:
I don't think you have to remove the class name here, as the previous loop that iterates over all the classnames already takes care of that (I think — did I miss smth?). This also means you can simplify the condition checks, too.

imalitavakoli replied:
Hmm... I will check it out and let you know :+1:

ingdir commented on jquery.bemhelpers.js:
```diff
@@ -62,40 +62,50 @@
-                
+
+                // if we're dealing with a String modifer(ONLY if we're dealing with a String modifer 'removeTheAlreadySetValue' may be true) and 'removeTheAlreadySetValue' is true, then it means that user aims to remove the already existing modifier value class.
+                if (removeTheAlreadySetValue === true) modVal = false;
```
I don't particularly like the idea of resetting the modVal to false here as it makes impossible for the event listeners to understand what has just happened (also, it's quite likely that whoever is listening to a key/value modifier change, always expects `modVal` to be a `String` instance, not a `Boolean`).

imalitavakoli replied:
Maybe when the modifier is removed, it would be good to set it to `undefined` instead of `false`. What do you think?

ingdir replied:
I'd say let's keep it consistent. If we make an empty string a special case (i.e. it removes the whole key-value modifier), let's also broadcast the same empty string back. This could also potentially make comparisons easier for the event listeners as we stick to the same data type.

ingdir commented on jquery.bemhelpers.js:
```diff
@@ -62,40 +62,50 @@
                 }
-                
+
+                // if we're dealing with a String modifer(ONLY if we're dealing with a String modifer 'removeTheAlreadySetValue' may be true) and 'removeTheAlreadySetValue' is true, then it means that user aims to remove the already existing modifier value class.
```
Could you split this (and similar long lines) into several lines for readability?

imalitavakoli committed:
Refactored PR #6

imalitavakoli commented:
- [x] I have refactored the codes a little bit.
- [x] Event handler returns an empty string for `modVal` if it has been removed.
- [x] Added test coverage.
- [x] Updated the documentation with more examples and clear explanation :+1:

ingdir commented:
@imalitavakoli Sorry for not responding earlier, I'll have some time to work on this in the upcoming days. Thank you!

ingdir commented:
Hey @imalitavakoli, thank you for the changes, LGTM now. I'll merge it, so your work becomes part of the master branch, but I will postpone releasing a new plugin version for another several days, as I'm still a bit unsure about the sequence of events that have to fire (or not fire) when a key/value modifier gets removed. I think your approach (triggering a wildcard event but not a specific one, and passing an empty string as a modifier val) is a reasonable choice and it will most likely stay like this.

Thank you for your contribution!
